Source/WebCore/svg/SVGGraphicsElement.cpp

```diff
@@ -148,10 +152,73 @@ void SVGGraphicsElement::svgAttributeChanged(const QualifiedName& attrName)
+                // update.
+                //
+                // There is no intrinsic reason for that, besides historical ones. If we decouple
+                // the 'font size screen scaling factor' from layout and only use it during painting
+                // we can optimize transformations for text, simply by avoid the need for layout.
+                auto previousTransform = transformFromLayer();
+                layerRenderer->updateLayerTransform();
+
+                auto currentTransform = transformFromLayer();
+                if (previousTransform == currentTransform)
+                    return false;
+
+                // Only if the effective x/y scale changes, a re-layout is necessary, due to changed on-screen scaling factors.
+                // The next RenderSVGText layout will see a different 'screen font scaling factor', different text metrics etc.
+                if (!WTF::areEssentiallyEqual(previousTransform.xScale(), currentTransform.xScale()))
+                    return true;
+
+                if (!WTF::areEssentiallyEqual(previousTransform.yScale(), currentTransform.yScale()))
+                    return true;
+
+                return false;
+            };
+
+            if (determineIfLayerTransformChangeModifiesScale()) {
+                if (auto* textAffectedByTransformChange = dynamicDowncast<RenderSVGText>(layerRenderer)) {
+                    // Mark text metrics for update, and only trigger a relayout and not an explicit repaint.
+                    textAffectedByTransformChange->setNeedsTextMetricsUpdate();
+                    textAffectedByTransformChange->textElement().updateSVGRendererForElementChange();
+                    return;
+                }
+
+                // Recursively mark text metrics for update in all descendant RenderSVGText objects.
+                bool markedAny = false;
+                for (auto& textDescendantAffectedByTransformChange : descendantsOfType<RenderSVGText>(*layerRenderer)) {
+                    textDescendantAffectedByTransformChange.setNeedsTextMetricsUpdate();
+                    textDescendantAffectedByTransformChange.textElement().updateSVGRendererForElementChange();
+                    if (!markedAny)
+                        markedAny = true;
+                }
+
+                // If we marked a text descendant for relayout, we are expecting a relayout ourselves, so no reason for an explicit repaint().
+                if (markedAny)
+                    return;
+            }
+
+            // Instead of performing a full-fledged layout (issuing repaints), just recompute the layer transform, and repaint.
+            // In LBSE transformations do not affect the layout (except for text, where it still does!) -- SVG follows closely the CSS/HTML route, to avoid costly layouts.
+            layerRenderer->updateLayerTransform();
+            layerRenderer->repaint();
```

Comment:
Issuing repaint from DOM feels a bit wrong to me. Did you consider moving most of this to the render tree?

Reply:
I guess you reference to the way repaints are issued when CSS properties change on e.g. HTML elements: style changes are handled in the render tree, and repaint is triggered from there, if necessary. However, changing e.g. SVGTransformList via SVG DOM, or the 'transform' _attribute_ (which is not mapped to the CSS property!) only results in a svgAttributeChanged() call in the DOM. That seems the only place where we could handle the transform update and issue repaints.

Hmm, or do you mean to just move the logic into rendering/ and call it from DOM? (That would be a good suggestion indeed :-)

Reply:
Right, I meant the latter, I know this can't be compared to how the style/CSS system handles things.

Reply:
Excellent, will do.